examples/OpticalFlow/flownet_models.py

```diff
@@ -0,0 +1,420 @@
+    dy = warp[:, 1, :, :]
+
+    xf = tf.reshape(tf.tile(tf.range(w), [h]), [h, w])
+    yf = tf.transpose(tf.reshape(tf.tile(tf.range(h), [w]), [w, h]), [1, 0])
+
+    xf = tf.cast(xf, dx.dtype)
+    yf = tf.cast(yf, dy.dtype)
+
+    xf = xf + dx
+    yf = yf + dy
+
+    alpha = tf.expand_dims(xf - tf.floor(xf), axis=0)
+    alpha = tf.expand_dims(xf - tf.floor(xf), axis=-1)
+    beta = tf.expand_dims(yf - tf.floor(yf), axis=0)
+    beta = tf.expand_dims(yf - tf.floor(yf), axis=-1)
+
+    xL = tf.clip_by_value(tf.cast(tf.floor(xf), dtype=tf.int32), 0, w - 1)
+    xR = tf.clip_by_value(tf.cast(tf.floor(xf) + 1, dtype=tf.int32), 0, w - 1)
+    yT = tf.clip_by_value(tf.cast(tf.floor(yf), dtype=tf.int32), 0, h - 1)
+    yB = tf.clip_by_value(tf.cast(tf.floor(yf) + 1, dtype=tf.int32), 0, h - 1)
+
+    batch_ids = tf.tile(tf.expand_dims(tf.expand_dims(tf.range(B), axis=-1), axis=-1), [1, h, w])
+
+    def get(y, x):
+        idx = tf.reshape(batch_ids * h * w + y * w + x, [-1])
+        idx = tf.cast(idx, tf.int32)
+        return tf.gather(img_flat, idx)
+
+    val = tf.zeros_like(alpha)
+    val += (1 - alpha) * (1 - beta) * tf.reshape(get(yT, xL), [-1, h, w, c])
+    val += (0 + alpha) * (1 - beta) * tf.reshape(get(yT, xR), [-1, h, w, c])
+    val += (1 - alpha) * (0 + beta) * tf.reshape(get(yB, xL), [-1, h, w, c])
+    val += (0 + alpha) * (0 + beta) * tf.reshape(get(yB, xR), [-1, h, w, c])
+
+    # we need to enforce the channel_dim known during compile-time here
+    shp = img.shape.as_list()
+    return tf.reshape(tf.transpose(val, [0, 3, 1, 2]), [-1, shp[1], h, w])
+
+
+def resize(x, factor=4, mode='bilinear'):
+    """Resize input tensor with unkown input-shape by a factor
+
+    Args:
+        x (tf.Tensor): tensor NCHW
+        factor (int, optional): resize factor for H, W
+
+    Note:
+        Differences here against Caffe have huge impacts on the
+        quality of the predictions.
+
+    Returns:
+        tf.Tensor: resized tensor NCHW
+    """
+    assert mode in ['bilinear', 'nearest']
+    shp = tf.shape(x)[2:] * factor
+    # NCHW -> NHWC
+    x = tf.transpose(x, [0, 2, 3, 1])
+    if mode == 'bilinear':
+        x = tf.image.resize_bilinear(x, shp, align_corners=True)
+    else:
+        # better approximation of what Caffe is doing
+        x = tf.image.resize_nearest_neighbor(x, shp, align_corners=False)
+    # NHWC -> NCHW
+    return tf.transpose(x, [0, 3, 1, 2])
+
+
+class FlowNetBase(ModelDesc):
+    def __init__(self, height=None, width=None, channels=3):
+        self.height = height
+        self.width = width
+        self.channels = channels
+
+    def inputs(self):
+        return [tf.placeholder(tf.float32, (1, self.channels, self.height, self.width), 'left'),
+                tf.placeholder(tf.float32, (1, self.channels, self.height, self.width), 'right')]
+
+    def graph_structure(self, x):
+        raise NotImplementedError()
+
+    def build_graph(self, left, right):
+        x = tf.stack([left, right], axis=2)
```

Comment:
It seems to me that all the other parts of the model uses the [2, C, H, W] layout. Why using a different layout here?

You can use [2, C, H, W] here to compute rgb_mean, and let `graph_structure` take [2, C, H, W]. I think this will remove many of the concat+stack

Reply:
Same response as above. I did a step-by-step debugging/comparison against Caffe. Changing these lines would destroy this validation script. But feel free to change it. 

My delivered version yields comparable AEE (with the tolerance one can expect between frameworks). If the AEE stays the same: I am ok.